lib/rules/prefer-named-capture-group.js

```diff
@@ -0,0 +1,120 @@
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+const {
+    CALL,
+    CONSTRUCT,
+    ReferenceTracker,
+    getStringIfConstant
+} = require("eslint-utils");
+const regexpp = require("regexpp");
+
+//------------------------------------------------------------------------------
+// Helpers
+//------------------------------------------------------------------------------
+
+const parser = new regexpp.RegExpParser();
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = {
+    meta: {
+        type: "suggestion",
+
+        docs: {
+            description: "enforce using named capture group in regular expression",
+            category: "Best Practices",
+            recommended: false,
+            url: "https://eslint.org/docs/rules/prefer-named-capture-group"
+        },
+
+        schema: [],
+
+        messages: {
+            required: "Capture group '{{group}}' in regular expression should be named."
+        }
+    },
+
+    create(context) {
+
+        /**
+         * Function to check regular expression.
+         *
+         * @param {string} regex The regular expression to be check.
+         * @param {ASTNode} node AST node which contains regular expression.
+         * @param {boolean} uFlag Flag indicates whether unicode mode is enabled or not.
+         * @returns {void}
+         */
+        function checkRegex(regex, node, uFlag) {
+            try {
+                const ast = parser.parsePattern(regex, 0, regex.length, uFlag);
+
+                regexpp.visitRegExpAST(ast, {
+                    onCapturingGroupEnter(group) {
+                        if (!group.name) {
+                            const locNode = node.type === "Literal" ? node : node.arguments[0];
+
+                            context.report({
+                                node,
+                                messageId: "required",
+                                loc: {
+                                    start: {
+                                        line: locNode.loc.start.line,
+                                        column: locNode.loc.start.column + group.start + 1
+                                    },
+                                    end: {
+                                        line: locNode.loc.start.line,
+                                        column: locNode.loc.start.column + group.end + 1
+                                    }
+                                },
+                                data: {
+                                    group: group.raw
+                                }
+                            });
+                        }
+                    }
+                });
+            } catch (_) {
```

Comment:
Can you move some code outside of the `try` block?

```js
let ast;
try {
    ast = parser.parsePattern(regex, 0, regex.length, uFlag);
} catch (_) {
    // ignore regex syntax errors
    return;
}

regexpp.visitRegExpAST(ast, {
  // ...
});
```

That way, if there is an unexpected error in the code in the `regexpp.visitRegExpAST` call (due to a bug), the error won't be ignored.